pkg/recovery/etcd.go

```diff
@@ -0,0 +1,147 @@
+// The etcd backend fetches control plane objects directly from etcd. This is adapted heavily from
+// kubernetes/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go.
+
+package recovery
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"reflect"
+	"strings"
+
+	"github.com/coreos/etcd/clientv3"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/conversion"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/pkg/api"
+)
+
+// EtcdBackend is a backend that extracts a controlPlane from an etcd instance.
+type EtcdBackend struct {
+	client     *clientv3.Client
+	decoder    runtime.Decoder
+	pathPrefix string
+}
+
+// NewEtcdBackend constructs a new EtcdBackend for the given client and pathPrefix.
+func NewEtcdBackend(client *clientv3.Client, pathPrefix string) Backend {
+	return &EtcdBackend{
+		client:     client,
+		decoder:    api.Codecs.UniversalDecoder(),
+		pathPrefix: pathPrefix,
+	}
+}
+
+// read implements Backend.read().
+func (s *EtcdBackend) read(ctx context.Context) (*controlPlane, error) {
+	cp := &controlPlane{}
+	for _, r := range []struct {
+		etcdName string
+		yamlName string
+		obj      runtime.Object
+	}{{
+		etcdName: "configmaps",
+		yamlName: "config-map",
+		obj:      &cp.configMaps,
+	}, {
+		etcdName: "daemonsets",
+		yamlName: "daemonset",
+		obj:      &cp.daemonSets,
+	}, {
+		etcdName: "deployments",
+		yamlName: "deployment",
+		obj:      &cp.deployments,
+	}, {
+		etcdName: "poddisruptionbudgets",
+		yamlName: "pod-disruption-budget",
+		obj:      &cp.podDisruptionBudgets,
+	}, {
+		etcdName: "secrets",
+		yamlName: "secret",
+		obj:      &cp.secrets,
+	}, {
+		etcdName: "services/specs",
+		yamlName: "service",
+		obj:      &cp.services,
+	}, {
+		etcdName: "serviceaccounts",
+		yamlName: "service-account",
+		obj:      &cp.serviceAccounts,
+	},
```

Comment:
Actually, looks to be slightly different behavior than I assumed.

On the one hand - the behavior of "extract everything" would enable stamping out cookie-cutter clusters from this source pretty easy (and this could be a nice feature). However, from a recovery standpoint, this feels pretty heavy-handed (why extract everything just to re-inject it where it already exists?).

From the purely recovery standpoint, I would think we only care about extracting enough to run a temporary control-plane - such that we get past the self-hosted chicken/egg issue and re-launch the cluster. All other assets will still exist and once the self-hosted control-plane is brought up - everything else should recover naturally.

The other situation we might want to address would be the potential for recovering from "bad" configuration that took down the control-plane (for whatever reason). This would mean we would need to extract some manifests to allow them to be modified, and then re-inject the fixed manifests into the cluster. However, I feel like this even could be scoped down to just core components (api, scheduler, controller-manager, proxy). This would also require a change to `bootkube start` such that it would replace existing assets (I believe `create` will error if object already exists).